sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -185,11 +187,26 @@ private[sql] class SharedState(
 }
 
 object SharedState extends Logging {
-  try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
-  } catch {
-    case e: Error =>
-      logWarning("URL.setURLStreamHandlerFactory failed to set FsUrlStreamHandlerFactory")
+  @volatile private var factory: Option[FsUrlStreamHandlerFactory] = None
+  private lazy val defaultFactory = new FsUrlStreamHandlerFactory()
+  private def setFsUrlStreamHandlerFactory(conf: SparkConf): Unit = {
+    factory match {
+      case Some(_) =>
+        logWarning("FsUrlStreamHandlerFactory has been already initialized, " +
+          "so it can not be modified")
+      case None => synchronized {
+        try {
+          if (conf.getBoolean("spark.fsUrlStreamHandlerFactory.enabled", true)) {
```

Comment:
Why do you need a config for this?

Reply:
This [SPARK-25694](https://issues.apache.org/jira/browse/SPARK-25694) is a long-standing issue. Originally, [[SPARK-12868][SQL] Allow adding jars from hdfs](https://github.com/apache/spark/pull/17342 ) added this for better Hive support. However, this have a side-effect when we use Spark without `-Phive`. This causes exceptions when the users tries to use another custom factories or 3rd party library (trying to set this). This configuration will unblock those non-hive users.

```
scala> sql("show tables").show
+--------+---------+-----------+
|database|tableName|isTemporary|
+--------+---------+-----------+
+--------+---------+-----------+


scala> java.net.URL.setURLStreamHandlerFactory(new org.apache.hadoop.fs.FsUrlStreamHandlerFactory())
java.lang.Error: factory already defined
  at java.net.URL.setURLStreamHandlerFactory(URL.java:1134)
  ... 47 elided
```

Reply:
I updated the PR description to give more context.

Reply:
It's order-sensitive. BTW, it seems that the configuration name should be revised.

Reply:
So it's there in case the user code needs to set this after, not before, this executes? Just trying to clarify if that's a valid use case

Reply:
Yes. Right, @srowen . And, the user needs to extend `org.apache.hadoop.fs.FsUrlStreamHandlerFactory`.

Reply:
@jiangzho It's very hard for end users to know this conf exists since it's hidden in the code, and not easy for developers to maintain this. In Spark community, we typically add a new conf with a default value and doc in `config/package.scala`.

For example, in this scenario, we can add

```scala
  private[spark] val USE_DEFAULT_URL_STREAM_HANDLER_FACTORY  =
    ConfigBuilder("spark.is.useDefaultUrlStreamHandlerFactory")
      .doc("Some doc")
      .booleanConf
      .createWithDefault(true)
```

Reply:
+1 for @dbtsai 's comment.

Reply:
+1 for @dbtsai 's comment.